hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/execution/datasources/lance/SparkLanceReaderBase.scala

```diff
@@ -0,0 +1,136 @@
+
+/**
+ * Reader for Lance files in Spark datasource.
+ * Implements vectorized reading using LanceArrowColumnVector.
+ *
+ * @param enableVectorizedReader whether to use vectorized reading (currently always true for Lance)
+ */
+class SparkLanceReaderBase(enableVectorizedReader: Boolean) extends SparkColumnarFileReader {
+
+  // Batch size for reading Lance files (number of rows per batch)
+  private val DEFAULT_BATCH_SIZE = 512
+
+  /**
+   * Read a Lance file with schema projection and partition column support.
+   *
+   * @param file              Lance file to read
+   * @param requiredSchema    desired output schema of the data (columns to read)
+   * @param partitionSchema   schema of the partition columns. Partition values will be appended to the end of every row
+   * @param internalSchemaOpt option of internal schema for schema.on.read (not currently used for Lance)
+   * @param filters           filters for data skipping. Not guaranteed to be used; the spark plan will also apply the filters.
+   * @param storageConf       the hadoop conf
+   * @return iterator of rows read from the file output type says [[InternalRow]] but could be [[ColumnarBatch]]
+   */
+  override def read(file: PartitionedFile,
+                    requiredSchema: StructType,
+                    partitionSchema: StructType,
+                    internalSchemaOpt: util.Option[InternalSchema],
+                    filters: Seq[Filter],
+                    storageConf: StorageConfiguration[Configuration],
+                    tableSchemaOpt: util.Option[MessageType] = util.Option.empty()): Iterator[InternalRow] = {
+
+    val filePath = file.filePath.toString
+
+    if (requiredSchema.isEmpty && partitionSchema.isEmpty) {
+      // No columns requested - return empty iterator
+      Iterator.empty
+    } else {
+      // Create child allocator for reading
+      val allocator = HoodieArrowAllocator.newChildAllocator(getClass.getSimpleName + "-data-" + filePath,
+        HoodieSparkLanceReader.LANCE_DATA_ALLOCATOR_SIZE);
+
+      try {
+        // Open Lance file reader
+        val lanceReader = LanceFileReader.open(filePath, allocator)
+
+        // Extract column names from required schema for projection
+        val columnNames: java.util.List[String] = if (requiredSchema.nonEmpty) {
+          requiredSchema.fields.map(_.name).toList.asJava
+        } else {
+          // If only partition columns requested, read minimal data
+          null
+        }
+
+        // Get schema from Lance file for HoodieLanceRecordIterator
+        val arrowSchema = lanceReader.schema()
+        val sparkSchema = LanceArrowUtils.fromArrowSchema(arrowSchema)
+
+        // Read data with column projection (filters not supported yet)
+        val arrowReader = lanceReader.readAll(columnNames, null, DEFAULT_BATCH_SIZE)
+
+        // Create iterator using shared HoodieLanceRecordIterator
+        val lanceIterator = new HoodieLanceRecordIterator(
```

Comment:
Does the iterator apply any required field schema evolution?

Reply:
@the-other-tim-brown was wondering if you can help me understand this better since I have a gap in my knowledge on the schema evolution piece. 

Currently the HoodieLanceRecordIterator is pretty much just the same iterator merged from the last reader PR https://github.com/apache/hudi/pull/17632/files#diff-c10a66239458bfdf557797999ca1c2616ddeb70e97898a40c5b10dbcf0dd172eR198, so it does not handle anything to do with schema evolution. Is the ask that we need to make changes to this iterator in order to support schema this and if so was wondering on what are some places in code base or general structure on how to approach this.

Otherwise was planning on following up on schema evolution in a separate task.